src/Mono.Android/Xamarin.Android.Net/NegotiateAuthenticationHelper.cs

```diff
@@ -0,0 +1,209 @@
+			HttpRequestMessage request,
+			RequestedNegotiateAuthenticationData requestedAuth,
+			CancellationToken cancellationToken)
+		{
+			using var authContext = new NegotiateAuthentication (
+				new NegotiateAuthenticationClientOptions {
+					Package = requestedAuth.AuthType,
+					Credential = requestedAuth.Credential,
+					TargetName = await GetTargetName (handler, request, requestedAuth.IsProxyAuth, cancellationToken).ConfigureAwait (false),
+					RequiredProtectionLevel = requestedAuth.IsProxyAuth ? ProtectionLevel.None : ProtectionLevel.Sign,
+				}
+			);
+
+			// we need to make sure that the handler doesn't override the authorization header
+			// with the user defined pre-authentication data
+			var originalPreAuthenticate = handler.PreAuthenticate;
+			handler.PreAuthenticate = false;
+
+			try {
+				return await DoSendWithAuthAsync (handler, request, authContext, requestedAuth, cancellationToken);
+			} finally {
+				handler.PreAuthenticate = originalPreAuthenticate;
+			}
+		}
+
+		static bool TryGetSupportedAuthType (string challenge, out string authType)
+		{
+			var spaceIndex = challenge.IndexOf (' ');
+			authType = spaceIndex == -1 ? challenge : challenge.Substring (0, spaceIndex);
+
+			return authType.Equals ("NTLM", StringComparison.OrdinalIgnoreCase) ||
+				authType.Equals ("Negotiate", StringComparison.OrdinalIgnoreCase);
+		}
+
+		static async Task <HttpResponseMessage?> DoSendWithAuthAsync (
+			AndroidMessageHandler handler,
+			HttpRequestMessage request,
+			NegotiateAuthentication authContext,
+			RequestedNegotiateAuthenticationData requestedAuth,
+			CancellationToken cancellationToken)
+		{
+			HttpResponseMessage? response = null;
+
+			int requestCounter = 0;
+			bool needDrain = false; // TODO maybe we need to drain even the first response on Android?
+			string? challengeData = null;
+
+			while (requestCounter++ < MaxRequests) {
+				var challengeResponse = authContext.GetOutgoingBlob (challengeData, out var statusCode);
+
+				if (challengeResponse is null || statusCode > NegotiateAuthenticationStatusCode.ContinueNeeded) {
+					// Response indicated denial even after login, so stop processing and return current response.
+					break;
+				}
+
+				if (needDrain) {
+					// We need to drain the content otherwise the next request
+					// won't reuse the same TCP socket and persistent auth won't work.
+					await response.Content.LoadIntoBufferAsync ().ConfigureAwait (false);
+				}
+
+				SetAuthorizationHeader (request, requestedAuth, challengeResponse);
+				response = await handler.DoSendAsync (request, cancellationToken).ConfigureAwait (false);
+
+				if (authContext.IsAuthenticated || !TryGetChallenge (response, requestedAuth, out challengeData)) {
+					break;
+				}
+
+				if (!IsAuthenticationChallenge (response, requestedAuth))
+				{
+					// Tail response for Negotiate on successful authentication. Validate it before we proceed.
+					authContext.GetOutgoingBlob(challengeData, out _);
```

Comment:
This needs to actually check the status code and throw an exception if it fails.